docs: add EFS to documentation

bvtujo opened the pull request:
<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bvtujo committed:
wip

bvtujo committed:
cdocs: add storage to documentation

bvtujo committed:
docs: add cross-links between sidecars and storage

bvtujo committed:
docs: rewrite storage concept page and add links from manifest

efekarakus commented on site/content/docs/developing/storage.md:
```diff
@@ -0,0 +1,219 @@
+# Storage
+
+There are two ways to add persistence to Copilot workloads: using [`copilot storage init`](#database-and-artifacts) to create databases and S3 buckets; and attaching an existing EFS filesystem using the [`storage` key](#file-systems) in the manifest. 
```
nit:
```suggestion
There are two ways to add persistence to Copilot workloads: using [`copilot storage init`](#database-and-artifacts) to create databases and S3 buckets; and attaching an existing EFS filesystem using the [`storage` field](#file-systems) in the manifest. 
```

efekarakus commented on site/content/docs/developing/storage.md:
````diff
@@ -0,0 +1,219 @@
+You can also create a [DynamoDB table](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Introduction.html) using `copilot storage init`. For example, to create the Cloudformation template for a table with a sort key and a local secondary index, you could run the following command.
+
+```bash
+$ copilot storage init -n users -t DynamoDB -w api --partition-key id:N --sort-key email:S --lsi post-count:N
````
I wonder if we should not show all these flags? To encourage users to use the guided path instead.

I saw in many of our issues that folks use the flags instead of prompts. I wonder if it's because of our examples or docs.

efekarakus replied:
Or maybe we can write it like this:
```suggestion
# For a guided experience.
$ copilot storage init

# Or by providing the flags to skip the prompts.
$ copilot storage init -n users -t DynamoDB -w api --partition-key id:N --sort-key email:S --lsi post-count:N
```

efekarakus commented on site/content/docs/developing/storage.md:
````diff
@@ -0,0 +1,219 @@
+  | jq -r '.Stacks[] | .Outputs[] | select(.OutputKey == "EnvironmentSecurityGroup") | .OutputValue')
+```
+
+Once you have these, creating mount targets is simple. 
````
nit: maybe we shouldn't say simple here :D since it's a lot of work to get to this point.
```suggestion
Once you have these, create the mount targets: 
```

efekarakus commented on site/content/docs/developing/storage.md:
````diff
@@ -0,0 +1,219 @@
+$ aws efs delete-mount-target --mount-target-id $MOUNT_TARGET_2
+```
+
+#### With Addons
````
I wonder if instead of addons, we should just say `CloudFormation` and not go through the hack of creating a job.

Something like:
```
If you'd like to create a EFS FS from scratch, you can use the create a new CloudFormation stack using the following template:
```

efekarakus commented on site/content/docs/developing/storage.md:
```diff
@@ -0,0 +1,219 @@
+storage:
+  volumes:
+    {{ volume name }}:
+      path: {{ path at which to mount }} # Required.
```
nit: should we clarify this is for the container? whereas `root_dir` below is clearly for the FS.

efekarakus commented on site/content/docs/developing/storage.md:
````diff
@@ -0,0 +1,219 @@
+```
+
+This will create an EFS file system and the mount targets needed to allow tasks to attach to it. You can get the ID of this filesystem with the AWS CLI and jq.
+```bash
+$ aws efs describe-file-systems | \
+  jq -r '.FileSystems[] | select((.Tags[]|select(.Key=="copilot-service")|.Value) =="efs-helper") | .FileSystemId'
+```
````
nit: instead of doing this can we `Output` the FileSystemID in the stack above so that in the yaml below we can just write:
```yaml
efs:
        id: {{ the FileSystemID output }}
```

efekarakus commented on site/content/docs/manifest/backend-service.md:
```diff
@@ -151,5 +151,10 @@ Key-value pairs that represent secret values from [AWS Systems Manager Parameter
 <div class="separator"></div>
 
+<a id="storage" href="#storage" class="field">`storage`</a> <span class="type">Map</span>  
+The Storage section lets you specify external EFS volumes for your containers and sidecars to mount. This allows you to access persistent storage across regions for data processing or CMS workloads. For more detail, see the [storage](../developing/storage.md) page.
```
nice, I think we should still include the subfields here though. These pages are a nice place for customers to see all the available fields in the manifest, on the other hand we don't need to go into as much detail.

efekarakus commented on site/content/docs/manifest/lb-web-service.md:
```diff
@@ -39,6 +39,19 @@ List of all available properties for a `'Load Balanced Web Service'` manifest.
     secrets:
       GITHUB_TOKEN: GITHUB_TOKEN
+
+    # Configuration for task storage using EFS volumes.
```
Can we remove this comment and instead move this blurb to the backend service right below the `count` field?  

I think it should be more common for folks to enable EFS on a backend service rather than lb web service.

efekarakus commented:
Great documentation btw ! 👍

Lou1415926 commented on site/content/docs/developing/storage.md:
````diff
@@ -0,0 +1,219 @@
+```
+The above command will create the Cloudformation template for an S3 bucket in the [addons](../developing/additional-aws-resources.md) directory for the "api" service. The next time you run `copilot deploy -n api`, the bucket will be created, permission to access it will be added to the `api` task role, and the name of the bucket will be injected into the `api` container under the environment vari ...
+
+All names are converted into SCREAMING_SNAKE_CASE based on their use of hyphens or underscores. You can view the environment variables for a given service by running `copilot svc show`.
````
A small question: should "All names....`copilot svc show`" be put into a !!!info block? 

I feel like this is like a small digress from the main topic "storage" and visually separating it somehow may help keeping the reading experience smooth.

Lou1415926 reviewed (commented):
Looks great! One small question regarding formatting.

bvtujo committed:
docs: address feedback

efekarakus commented on site/content/docs/developing/storage.md:
````diff
@@ -0,0 +1,213 @@
+```
+
+#### CloudFormation
+Here's an example of how you might create the appropriate EFS infrastructure for an external file system using the [Addons](../developing/additional-aws-resources.md) functionality. 
````
We can remove this line now :)

efekarakus commented on site/content/docs/manifest/backend-service.md:
```diff
@@ -23,6 +23,20 @@ List of all available properties for a `'Backend Service'` manifest.
     memory: 512
     count: 1
+
+    # Configuration for task storage using EFS volumes.
```
nit: can we delete this line?

efekarakus reviewed (approved):
Feel free to remove do-not-merge after addressing!

huanjani commented on site/content/docs/developing/storage.md:
```diff
@@ -0,0 +1,213 @@
+## File Systems
+Mounting an EFS volume in Copilot tasks requires two things:
+
+1. That you create an [EFS file system](https://docs.aws.amazon.com/efs/latest/ug/whatisefs.html) in the region of the environment you wish to use it with
```
nit: add period to match below? or add "or" to the end?
And maybe "...in the desired environment's region" or "...in the target environment's region" or something like that to make it less wordy?

huanjani commented on site/content/docs/developing/storage.md:
````diff
@@ -0,0 +1,213 @@
+```yaml
+storage:
+  volumes:
+    myEFSVolume: # This is a variable key and can be set to arbitrary strings.
````
nit: 
```suggestion
    myEFSVolume: # This is a variable key and can be set to an arbitrary string.
```

huanjani commented on site/content/docs/developing/storage.md:
````diff
@@ -0,0 +1,213 @@
+        id: fs-1234567 
+```
+
+This will create a read-only mounted volume in your service or job's container using the filesystem `fs-1234567`. If mount targets are not created in the subnets of the environment, the task will fail to launch. 
````
nit:
```suggestion
This will create a read-only mounted volume in your service's or job's container using the filesystem `fs-1234567`. If mount targets are not created in the subnets of the environment, the task will fail to launch. 
```

huanjani reviewed (approved):
This looks fantastic! Just a few tiny nits.

bvtujo committed:
docs: address feedback

Rewrite storage guide to use manual `aws cloudformation` commands instead of the arguably hackier Addons method.

bvtujo committed:
Merge branch 'mainline' into efs-docs
